public/requests/queries/dashboard_queries.ts

```diff
@@ -31,39 +31,24 @@ export const getDashboardQuery = () => {
           size: 10000,
         },
         aggs: {
-          traces: {
-            terms: {
-              field: 'traceId',
-              order: {
-                last_updated: 'desc',
-              },
-              size: 10000,
-            },
-            aggs: {
-              duration: {
-                max: {
-                  field: 'traceGroupFields.durationInNanos',
-                },
-              },
-              last_updated: {
-                max: {
-                  field: 'traceGroupFields.endTime',
-                },
-              },
-            },
-          },
-          average_latency_nanos: {
-            avg_bucket: {
-              buckets_path: 'traces>duration',
-            },
-          },
           average_latency: {
-            bucket_script: {
-              buckets_path: {
-                count: '_count',
-                latency: 'average_latency_nanos.value',
-              },
-              script: 'Math.round(params.latency / 10000) / 100.0',
+            scripted_metric: {
+              init_script: 'state.map = [:]; state.total_latency = 0.0; state.count = 0.0;',
+              map_script: `
+                if (doc.containsKey('traceGroupFields.durationInNanos') && !doc['traceGroupFields.durationInNanos'].empty) {
+                  def id = doc['traceId'].value;
+                  if (!state.map.containsKey(id)) {
+                    state.map[id] = true;
+                    state.count++;
+                    state.total_latency += doc['traceGroupFields.durationInNanos'].value;
+                  }
+                }
+                `,
+              combine_script: `
+                def average_latency_nanos = state.total_latency / state.count;
+                return Math.round(average_latency_nanos / 10000) / 100.0;
+                `,
+              reduce_script: 'return states[0]',
```

Comment:
It seems like the averaging logic in `combine_script` should be moved down to `reduce_script`. I'm just getting familiar with the [scripted metric docs](https://www.elastic.co/guide/en/elasticsearch/reference/7.10/search-aggregations-metrics-scripted-metric-aggregation.html), but these points stand out:

* combine_script: Executed **once on each shard** after document collection is complete. Allows the aggregation to consolidate the state returned from each shard.
* reduce_script: Executed **once on the coordinating node** after all shards have returned their results. The script is provided with access to a variable `states` which is an array of the result of the combine_script on each shard.

So with this in mind, wouldn't the reduce_script need to iterate over the states[] array, using the `total_latency` and `count` values returned by each shard (instead of just taking `states[0]`)?  I've seen a few examples where `combine_script` is just `"return state"`.

Reply:
I originally have the reduce script to be 
```
              combine_script: `
                return state.total_latency / state.count;
              `,
              reduce_script: `
                def total = 0.0;
                for (latency_nanos in states) total += latency_nanos;
                return Math.round(total / states.length / 10000) / 100.0;
              `,
```
It works on the main dashboard query, but on latency trends query the same scripted metrics agg is giving me `search_phase_execution_exception`.
```
      "caused_by" : {
        "type" : "null_pointer_exception",
        "reason" : "Cannot invoke \"Object.getClass()\" because \"right\" is null"
      }
```
Appreciate if you could help...